packages/@ember/-internals/glimmer/lib/modifiers/on.ts

```diff
@@ -0,0 +1,241 @@
+import { assert } from '@ember/debug';
+import { DEBUG } from '@glimmer/env';
+import { Opaque, Simple } from '@glimmer/interfaces';
+import { Tag } from '@glimmer/reference';
+import { Arguments, CapturedArguments, ModifierManager } from '@glimmer/runtime';
+import { Destroyable } from '@glimmer/util';
+
+/*
+  Internet Explorer 11 does not support `once` and also does not support
+  passing `eventOptions`. In some situations it then throws a weird script
+  error, like:
+
+  ```
+  Could not complete the operation due to error 80020101
+  ```
+
+  This flag determines, whether `{ once: true }` and thus also event options in
+  general are supported.
+*/
+const SUPPORTS_EVENT_OPTIONS = (() => {
+  try {
+    const div = document.createElement('div');
+    let counter = 0;
+    div.addEventListener('click', () => counter++, { once: true });
+
+    let event;
+    if (typeof Event === 'function') {
+      event = new Event('click');
+    } else {
+      event = document.createEvent('Event');
+      event.initEvent('click', true, true);
+    }
+
+    div.dispatchEvent(event);
+    div.dispatchEvent(event);
+
+    return counter === 1;
+  } catch (error) {
+    return false;
+  }
+})();
+
+export class OnModifierState {
+  public tag: Tag;
+  public element: Element;
+  public args: CapturedArguments;
+  public eventName!: string;
+  public callback!: EventListener;
+  private userProvidedCallback!: EventListener;
+  public once?: boolean;
+  public passive?: boolean;
+  public capture?: boolean;
+  public options?: AddEventListenerOptions;
+  public shouldUpdate = true;
+
+  constructor(element: Element, args: CapturedArguments) {
+    this.element = element;
+    this.args = args;
+    this.tag = args.tag;
+  }
+
+  updateFromArgs() {
+    let { args } = this;
+
+    let { once, passive, capture }: AddEventListenerOptions = args.named.value();
+    if (once !== this.once) {
+      this.once = once;
+      this.shouldUpdate = true;
+    }
+
+    if (passive !== this.passive) {
+      this.passive = passive;
+      this.shouldUpdate = true;
+    }
+
+    if (capture !== this.capture) {
+      this.capture = capture;
+      this.shouldUpdate = true;
+    }
+
+    let options: AddEventListenerOptions;
+    if (once || passive || capture) {
+      options = this.options = { once, passive, capture };
+    } else {
+      this.options = undefined;
+    }
+
+    assert(
+      'You must pass a valid DOM event name as the first argument to the `on` modifier',
+      args.positional.at(0) !== undefined && typeof args.positional.at(0).value() === 'string'
```

Comment:
Similar to asserting that only `once`, `passive` and `capture` are passed as named arguments, we could also check that only 2 positional arguments are passed.